### PR TITLE
feat: throw friendly warning message for ambiguous export

### LINKED ISSUE #1621: Throw friendly warning message for ambiguous export
### What problem does this feature solve?

.

### What does the proposed API look like?

- Reference https://github.com/rolldown/rolldown/pull/397 for what the error message looks like
- https://github.com/rolldown/rolldown/blob/24078fb0d2e060d5c51f6c5d82b569139d886320/crates/rolldown_error/src/build_error/error_constructors.rs
- Where you need to report the warning
https://github.com/rolldown/rolldown/blob/24078fb0d2e060d5c51f6c5d82b569139d886320/crates/rolldown/src/stages/link_stage/bind_imports_and_exports.rs#L270

### PR DESCRIPTION
**Pull request by shulaoda**
Closes #1621.
Closes #397.

### CODE REVIEW

**Comment by IWANABETHATGUY** [comment]
Can you fix all compilation errors before we can push forward?

File: crates/rolldown_error/src/events/ambiguous_external_namespace.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  fn kind(&self) -> EventKind {
+    EventKind::AmbiguousExternalNamespace
+  }
+
+  fn message(&self, _opts: &DiagnosticOptions) -> String {
+    format!(
+      r#""{}" re-exports "{}" from one of the modules {} (will be ignored)."#,
+      self.importer,
+      self.imported_specifier,
+      self.importee.iter().map(|v| format!(r#""{v}""#)).collect::<Vec<_>>().join(" and ")
```
Comment (IWANABETHATGUY) [inline crates/rolldown_error/src/events/ambiguous_external_namespace.rs]: You need to consider the following scenarios: 
1. the  ambiguous **re-export** exists in two modules.
2. the ambiguous **re-export** exists in more than two modules.
## Related  Grammar
Using a List with Commas and a Conjunction
The most common way to list items is by using commas to separate each item and using "and" (or "or") before the last item.

Example:

We need to buy apples, oranges, and bananas.

File: crates/rolldown/tests/fixtures/errors/ambiguous_external_namespace/artifacts.snap [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+assertion_line: 176
+expression: content
+input_file: crates/rolldown/tests/fixtures/errors/ambiguous_external_namespace
+---
+# Errors
+
+## AMBIGUOUS_EXTERNAL_NAMESPACES
+
+```text
+[AMBIGUOUS_EXTERNAL_NAMESPACES] Error: Warning: Found ambiguous export.
```
Comment (IWANABETHATGUY) [inline crates/rolldown/tests/fixtures/errors/ambiguous_external_namespace/artifacts.snap]: The diagnostic is incorrect, you should adjust `Severity` and remove `Warning: ` in the content.

File: crates/rolldown_error/src/events/ambiguous_external_namespace.rs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    )
+  }
+
+  fn on_diagnostic(&self, diagnostic: &mut Diagnostic, opts: &DiagnosticOptions) {
+    let file_id =
+      diagnostic.add_file(self.importer_filename.clone(), Arc::clone(&self.importer_source));
+
+    diagnostic.title = "Warning: Found ambiguous export.".to_string();
```
Comment (IWANABETHATGUY) [inline crates/rolldown_error/src/events/ambiguous_external_namespace.rs]:
```suggestion
    diagnostic.title = "Found ambiguous export.".to_string();
    diagnostic.severity = Severity::Warning;
```

File: crates/rolldown_error/src/events/ambiguous_external_namespace.rs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      self.imported_specifier,
+      self.importee.iter().map(|v| format!(r#""{v}""#)).collect::<Vec<_>>().join(" and ")
+    )
+  }
+
+  fn on_diagnostic(&self, diagnostic: &mut Diagnostic, opts: &DiagnosticOptions) {
+    let file_id = diagnostic.add_file(self.importer_filename.clone(), self.importer_source.clone());
+
+    diagnostic.title = "Found ambiguous export.".to_string();
+    diagnostic.severity = Severity::Warning;
```
Comment (hyf0) [inline crates/rolldown_error/src/events/ambiguous_external_namespace.rs]: Don't need to specify severity here.

See

https://github.com/rolldown/rolldown/blob/65b119ab4c42584735738288a6c3705b16e854e1/crates/rolldown/src/ast_scanner/mod.rs#L473-L481

File: crates/rolldown_error/src/events/ambiguous_external_namespace.rs (hunk 1 above)
Comment (shulaoda) [reply crates/rolldown_error/src/events/ambiguous_external_namespace.rs]: Yes, I have considered it. Is it like below, I just need to optimize the separation?

https://github.com/rolldown/rolldown/blob/06ca7c787047e225e71524af5b2d6c3038eef0f4/crates/rolldown/tests/fixtures/errors/ambiguous_external_namespace/artifacts.snap#L15-L17

Comment (IWANABETHATGUY) [reply crates/rolldown_error/src/events/ambiguous_external_namespace.rs]: that's not correct, when there is ambiguous **re-export** from three different modules, the message should be 
```bash
"a.js", "b.js" and "c.js"
```

Comment (IWANABETHATGUY) [reply crates/rolldown_error/src/events/ambiguous_external_namespace.rs]: for two modules, it should be : 
```bash
"a.js" and "b.js"
```

File: crates/rolldown/tests/fixtures/errors/ambiguous_external_namespace/_config.json [hunk 5]
```diff
@@ -0,0 +1,3 @@
+{
+  "expectError": true
```
Comment (hyf0) [inline crates/rolldown/tests/fixtures/errors/ambiguous_external_namespace/_config.json]: Not needed for emitting warnings.

File: crates/rolldown/src/stages/link_stage/bind_imports_and_exports.rs [hunk 6]
```diff
@@ -265,7 +264,34 @@ impl<'a> BindImportsAndExportsContext<'a> {
...
+          importee.extend(
+            potentially_ambiguous_symbol_refs
+              .iter()
+              .map(|&symbol_ref| {
+                self.normal_modules[symbol_ref.owner].stable_resource_id().to_string()
+              })
+              .collect::<Vec<_>>(),
+          );
+
+          self.errors.push(
```
Comment (hyf0) [inline crates/rolldown/src/stages/link_stage/bind_imports_and_exports.rs]: Add a `self.warnings` to store warnings.

**Comment by hyf0** [comment]
This is fine to be merged. Could you run `just test-rust` and update all changed snapshots by `cargo insta accept`.

**Comment by IWANABETHATGUY** [comment]
```diff
diff --git a/crates/rolldown/tests/esbuild/dce/disable_tree_shaking/_config.json b/crates/rolldown/tests/esbuild/dce/disable_tree_shaking/_config.json
index cc39d9bb..94d0fe2d 100644
--- a/crates/rolldown/tests/esbuild/dce/disable_tree_shaking/_config.json
+++ b/crates/rolldown/tests/esbuild/dce/disable_tree_shaking/_config.json
@@ -8,7 +8,7 @@
     ],
     "treeshake": false,
     "external": [
-      "react"
+      "react/jsx-runtime"
     ]
   },
   "expectExecuted": false
```
to avoid changes of **crates/rolldown/tests/esbuild/dce/disable_tree_shaking/artifacts.snap**

File: crates/rolldown/tests/fixtures/warnings/ambiguous_external_namespace/_config.json [hunk 7]
```diff
@@ -0,0 +1,3 @@
+{
+  "expectError": false
```
Comment (hyf0) [inline crates/rolldown/tests/fixtures/warnings/ambiguous_external_namespace/_config.json]:
```suggestion
  "expectExecuted": false
```

File: crates/rolldown/tests/fixtures/warnings/ambiguous_external_namespace/artifacts.snap [hunk 8]
```diff
@@ -0,0 +1,31 @@
+---
+source: crates/rolldown_testing/src/case/case.rs
+assertion_line: 176
+expression: content
+input_file: crates/rolldown/tests/fixtures/warnings/ambiguous_external_namespace
+---
+# warnings
+
+## AMBIGUOUS_EXTERNAL_NAMESPACES
```
Comment (hyf0) [inline crates/rolldown/tests/fixtures/warnings/ambiguous_external_namespace/artifacts.snap:9]: https://hyrious.me/esbuild-repl/?version=0.23.0&b=e%00entry.js%00import+%7B+foo+%7D+from+%27.%2Fproxy%27%0Aconsole.log%28foo%29&b=%00a.js%00export+const+foo+%3D+1&b=%00b.js%00export+const+foo+%3D+2&b=%00proxy.js%00export+*+from+%27.%2Fa%27%0Aexport+*+from+%27.%2Fb%27&o=--bundle+--splitting+--format%3Desm+--external%3Aext+--outdir%3Ddist

Esbuild treats this as an error and has better report. Let's merge this PR for now. I'll open an new issue to port esbuild's behavior.

**Review by hyf0 — APPROVED** [review]
Thanks

**Pull request merged**